### ISSUE TITLE
Auto-colour Projektzeit entries from the selected Projekt and Vorgang

### ISSUE THREAD

**Issue by ChristophRettinger** [ChristophRettinger/ASC.ZepExtension#1]
## Problem Statement

ZEP lets a Projektzeit carry a Farbe, chosen from ten swatches in the "Projektzeit
erfassen" dialog, and a coloured week view is markedly easier to read at a glance than an
uncoloured one. But the colour has to be clicked by hand on every single entry, and it is
in no way connected to the Projekt the time is booked to. In practice that means either
colouring every entry manually — tedious enough that it stops happening — or giving up on
the feature and reading a wall of identical white blocks.

What is wanted is for the colour to follow from the Projekt automatically: the same
Projekt should always look the same, without anyone deciding that per entry.

## Solution

A userscript that watches the Projektzeit dialog and sets the Farbe from the selected
Projekt and Vorgang, according to a configurable list of Rules.

When a new Projektzeit dialog opens, or whenever the Projekt or Vorgang selection changes,
the script finds the first matching Rule and selects the corresponding palette slot. If no
Rule matches, it leaves the Farbe alone. It never touches an existing Projektzeit that was
merely opened for editing.

Rules are configured in a dialog injected into the ZEP page itself, editable as rows —
Projekt picked from the live dropdown, colour picked from the real swatches — and stored
locally, layered over a set of defaults that ships in the script file.

## User Stories

1. As a ZEP user, I want a new Projektzeit to be coloured automatically from its
   pre-selected Projekt, so that entries I create without changing anything are already
   colour-coded.
2. As a ZEP user, I want the Farbe to update when I pick a different Projekt in the
   dialog, so that the colour always reflects what I actually booked to.
3. As a ZEP user, I want the Farbe to update when I pick a different Vorgang, so that
   sub-divisions of one Projekt can be distinguished where that matters.
4. As a ZEP user, I want a Projekt with no matching Rule to be left uncoloured, so that a
   missing Rule is visible rather than silently guessed at.
5. As a ZEP user, I want to be able to click a swatch myself after choosing the Projekt,
   so that I can override the automatic colour for a one-off entry.
6. As a ZEP user, I want my manual override to survive, so that the script does not undo a
   deliberate choice a moment after I make it.
7. As a ZEP user, I want opening an old Projektzeit to leave its Farbe untouched, so that
   colours I set by hand months ago are never silently rewritten.
8. As a ZEP user, I want the script to work when I create an entry from a template or work
   package, so that the colouring is consistent across every route into the dialog.
9. As a ZEP user, I want a Rule to be able to match a family of Projekte by prefix, so that
   one line covers "ASC intern API", "ASC intern DB" and everything else in that family.
10. As a ZEP user, I want Rules evaluated in order with the first match winning, so that I
    can put a narrow exception above a broad family rule and have it take precedence.
11. As a ZEP user, I want a Rule to optionally narrow to a specific Vorgang, so that one
    Projekt's support work can look different from its development work.
12. As a ZEP user, I want Rules for Projekte without a Vorgang to work unchanged, so that
    the many Projekte that have no Vorgang are not a special case I have to think about.
13. As a ZEP user, I want to open a configuration dialog from inside ZEP, so that I do not
    have to edit a script file to change a colour.
14. As a ZEP user, I want to add a Rule as a row, so that configuration does not require
    knowing any syntax.
15. As a ZEP user, I want to pick the Projekt for a Rule from the live dropdown, so that I
    cannot mistype a name and produce a Rule that silently never matches.
16. As a ZEP user, I want the picked Projekt name to remain editable text, so that I can
    trim a full name down to the prefix that covers a family.
17. As a ZEP user, I want to pick a Rule's colour from the actual ZEP swatches, so that I
    can see what the entry will look like rather than reading a hex code.
18. As a ZEP user, I want to reorder Rules, so that I can control which one wins when two
    could match.
19. As a ZEP user, I want to delete a Rule, so that retired projects do not accumulate.
20. As a ZEP user, I want my configuration to persist across sessions and page loads, so
    that I set it up once.
21. As a ZEP user, I want a warning on any Rule that matches none of the Projekte currently
    on offer, so that a renamed or finished project shows up as a stale Rule.
22. As a ZEP user, I want to see which Rule matched for the dialog that is open, so that I
    can tell "no Rule matched" apart from "the script is broken".
23. As a ZEP user, I want to view and edit my Rules as JSON, so that I can make bulk changes
    quickly.
24. As a ZEP user, I want to export my Rules, so that I can hand a working set to a
    colleague or paste it into the script's defaults.
25. As a ZEP user, I want to import a set of Rules, so that I can adopt a colleague's setup
    without retyping it.
26. As a colleague, I want to install the script from a single URL, so that trying it out
    costs me nothing.
27. As a colleague, I want sensible colours out of the box without configuring anything, so
    that the script is useful the moment it is installed.
28. As a colleague, I want my own local changes to override the shipped defaults, so that I
    can deviate without editing the shared file.
29. As a colleague, I want the script to update itself when a new version is pushed, so that
    I am not stuck on an old copy.
30. As a colleague at a different ZEP tenant, I want the script to work without editing the
    match pattern, so that it is not hard-wired to one company.
31. As a maintainer, I want debug tracing behind a flag, so that I can diagnose a failure
    without shipping noise to everyone's console.
32. As a maintainer, I want the script to fail quietly and leave ZEP fully usable if its
    assumptions break, so that a ZEP upgrade can never block anyone from booking time.
33. As a maintainer, I want the colour-resolution logic to be unit-testable without a
    browser, so that the rule-matching edge cases are covered by fast tests.

## Implementation Decisions

**Delivery.** A single userscript, per ADR 0001. `@grant none` so it runs in page context;
`@match https://www.zep-online.de/*/view/*` so any ZEP tenant is covered and no session
parameter appears in the pattern. `@updateURL`/`@downloadURL` point at the raw file on the
public repo's default branch, giving colleagues auto-update.

**Event binding goes through the page's jQuery.** The Projekt and Vorgang controls are real
`<select>` elements (`#projektId`, `#vorgangId`) enhanced by select2. select2 emits *jQuery*
change events, which a native `addEventListener('change')` does not receive. All binding
therefore uses `window.jQuery`, delegated from `document`, so handlers survive the dialog
being destroyed and rebuilt.

**Dialog lifecycle hook.** The dialog is a Bootstrap modal on a single reused element,
`#zep-popup`. Opening is detected via `shown.bs.modal`; no `MutationObserver` is needed.

**Writing the Farbe is a plain radio selection.** The swatches are ten
`input[type=radio][name=color]` elements whose value is the literal hex. `ProjektzeitFormMgr.js`
contains no reference to `color`, so the radios carry no JavaScript behaviour and no change
event needs synthesizing — setting `checked` is sufficient and is what gets posted.

**The palette is closed at ten values**, named once in a `PALETTE` table so that Rules refer
to colours by name rather than by hex or by position:

```
none   #e6f2f0   (default)      lilac  #EADFF7
sage   #b5d8d2                  blush  #FFE6E0
cream  #FFF2D9                  amber  #FFD173   (saturated)
mint   #DEF6E7                  green  #86DFA7   (saturated)
ice    #D9F4F9                  cyan   #73D8EA   (saturated)
```

**Rule shape.** An ordered list, first match wins. Each Rule matches the *visible label* of
the Projekt by prefix — not the option's id — so that configuration stays human-editable and
shareable, and so that a rename produces a visibly uncoloured entry rather than a silently
wrong colour. A Rule may additionally narrow to a Vorgang label. Matching Tätigkeit is not
supported; it is unused at this company.

**Configuration is layered.** A `DEFAULT_RULES` constant in the script file is the shared,
diffable baseline; `localStorage` holds only the local deviations and is written by the
injected editor. A colleague who never opens the editor gets the defaults.

**Unmapped Projekte are left untouched** — no fallback colour, and specifically no hashing of
the Projekt name onto a slot, because ten slots guarantee collisions between unrelated
projects.

**Write timing**, per ADR 0002: on new-dialog-open and on every Projekt/Vorgang change, never
on open of an existing entry. No "human set this" flag is stored; because the script only
writes at change time, a later manual click is inherently the last write.

**New and edit dialogs are distinguished structurally, not heuristically.** ZEP names the
controls differently in the two modes. The new-entry dialog uses bare ids — `#projektId`,
`#vorgangId`, `#taetigkeit`, `#von`/`#bis`/`#dauer` — while an entry opened for editing uses
`popup_`-prefixed ones: `#popup_projektId`, `#popup_vorgangId`, `#popup_taetigkeit`,
`#popupvon`/`#popupbis`/`#popupdauer`. Fields such as `color`, `datum`, `fakturierbar`,
`bemerkung` and `ort` keep the same name in both. The modal title also differs
("Projektzeit erfassen" versus "Projektzeit ändern"), but it is locale-dependent and is not
relied upon.

Binding exclusively to the bare ids therefore delivers ADR 0002's behaviour by construction:
the script is structurally incapable of seeing an edit dialog, and no runtime "is this a new
entry?" check exists to get wrong. Using the checked colour as a discriminator is rejected
outright, because an uncoloured existing entry is indistinguishable from a new one that way.

This also means the `copyVonVorlage` / `copyVonAGP` / `copyVonArbeitspaket` creation routes
need no special handling as long as they populate the bare-id form; that should be confirmed
the first time one of them is used.

**Status reporting.** The floating button that opens the editor doubles as a status light,
reporting the matched Rule or "no rule" for the dialog currently open. Tracing beyond that
sits behind a `DEBUG` flag.

**Failure posture.** Every hook is defensive: if a selector or global is missing, the script
logs under `DEBUG` and does nothing. Booking time in ZEP must never be blocked by this script.

## Testing Decisions

A good test here exercises external behaviour — given a set of Rules and a
(Projekt, Vorgang) pair, which palette slot comes out — and asserts nothing about how the
matching is implemented internally.

**One seam, as high as possible: the pure colour resolution.** The rule-matching and
configuration-merging logic is written as pure functions over plain data, with no DOM and no
jQuery. That is where every logic bug lives: prefix matching, ordering and first-match-wins,
Vorgang narrowing, the unmapped case, and merging `localStorage` overrides onto
`DEFAULT_RULES`. These are covered by `node:test`, which needs no dependencies.

To keep this testable without introducing a build step — which would defeat the fast
edit-and-reload loop that motivated choosing a userscript — the script exports its pure
functions under a `typeof module !== 'undefined'` guard, and guards its DOM bootstrap behind
a `typeof window !== 'undefined'` check. Node can then require the `.user.js` file directly.

**The DOM adapter is deliberately not unit-tested.** Binding to select2, reading the selected
option label and checking a radio are thin glue over a third-party application whose markup
we do not control; a mock of that markup would assert our assumptions rather than ZEP's
behaviour, and would keep passing on the day ZEP changes. It is verified by hand against the
live application, which is what the `DEBUG` flag and the status light exist to support.

There is no prior art in this repository: it is empty apart from the domain documentation.

## Out of Scope

- **Colouring the existing backlog.** Months of uncoloured entries stay uncoloured. Fixing
  them means either driving the dialog open-and-save per entry, which is slow and churns
  ZEP's change history, or bulk-PATCHing through the ZEP REST API — a separate one-off script
  with a bearer token, not a userscript.
- **Tätigkeit as an input to the colour decision.** It is unused at this company.
- **ZEP's `next/` frontend.** The `view/` interface is the one in use.
- **Packaging as a signed Firefox extension**, per ADR 0001.
- **Any server-side or shared configuration.** Rules live in each person's browser.
- **Arbitrary colours.** ZEP offers ten and only ten.
- **Changing anything about how ZEP renders colours** in the week view or in reports.

## Further Notes

The target instance is ZEP v7.13 at `www.zep-online.de/zepaschauer/`. The stack behind the
`view/` interface is server-rendered PHP with jQuery 2.2.4, select2, Bootstrap modals,
FullCalendar for the week view, and ZEP's own helper layer (`zepSelectBox`, `zepPopup`,
`dispatchZepEvent`, which is simply `window.dispatchEvent(new Event(type))`). The dialog form
posts to `ajax.php`.

ZEP carries its session in the URL as a `CLIENTSESSID` query parameter. That value is a live
credential: it must never be logged, committed, or included in a bug report.

The palette is not uniform. Six slots are pale pastels, three (`amber`, `green`, `cyan`) are
saturated and genuinely leap out of a week view, and one is the near-white default. Rules
should spend the saturated three on the projects that most need to be spotted at a glance.

The Projekt dropdown in the dialog offered only three options during reconnaissance. Whether
that is the full list of active Projekte or a filtered subset is unconfirmed, and it bounds
how much the row editor's dropdown can offer.

The `#zep-popup` modal is `position: fixed`, so its `offsetParent` is `null` even while it is
fully visible. Visibility must be tested via the `show` class, never via `offsetParent`.
